File: cli/src/main/java/io/bitken/ss/cli/plan/Preflight.java
package io.bitken.ss.cli.plan;

import io.bitken.ss.cli.HasSpec;
import io.bitken.ss.gw.GitState;
import io.bitken.ss.gw.GitTags;
import picocli.CommandLine.Model.CommandSpec;
import picocli.CommandLine.Model.OptionSpec;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

/**
 * {@code plan preflight --plan N} — four-condition verifier replacing the
 * step-6 bash block in phase1-plan.
 *
 * <p>FAIL (non-zero exit): dirty working tree, version tag absent locally.
 * <p>WARN (pass overall): branch not pushed or HEAD ahead of upstream.
 */
public class Preflight implements Callable<Integer>, HasSpec {

    private final CommandSpec spec;
    private final GitState gitState;
    private final GitTags gitTags;

    public Preflight(GitState gitState, GitTags gitTags) {
        this.gitState = gitState;
        this.gitTags = gitTags;
        spec = CommandSpec.wrapWithoutInspection(this);
        spec.name("preflight");
        spec.usageMessage().description("Verify plan preconditions before Phase 2.");
        spec.addOption(OptionSpec.builder("--plan").paramLabel("PLAN_NUMBER").required(true).description("Plan number").type(int.class).build());
    }

    @Override
    public CommandSpec getSpec() {
        return spec;
    }

    @Override
    public Integer call() {
        var pr = spec.commandLine().getParseResult();
        int plan = pr.matchedOption("plan").getValue();

        List<String> warnings = new ArrayList<>();

        if (!gitState.isClean()) {
            System.out.println("FAIL: working tree has uncommitted changes (git status --porcelain)");
            return 1;
        }

        String versionTag = gitTags.getPlanVersion(plan);
        if (!gitState.tagExistsLocally(versionTag)) {
            System.out.println("FAIL: version tag " + versionTag + " not found locally");
            return 1;
        }

        if (!gitState.isBranchPushedAndNotAhead()) {
            warnings.add("WARN: branch is not pushed or HEAD is ahead of upstream");
        }

        if (!gitState.tagExistsOnRemote(versionTag)) {
            warnings.add("WARN: version tag " + versionTag + " not found on remote");
        }

        warnings.forEach(System.out::println);
        System.out.println("PASS");
        return 0;
    }
}

File: exp/rust/crates/ss-cli/src/plan/preflight.rs
//! `plan preflight --plan N` — the four-condition verifier that replaced the
//! step-6 bash block in the plan skill.
//!
//! Port of the Java `Preflight`. The FAIL/WARN split is the contract: a dirty
//! tree or a version tag missing locally returns 1 **immediately** (later
//! conditions are never evaluated), while an unpushed branch and a tag absent
//! from the remote are warnings printed before the final `PASS`.

use ss_core::gw::{GitState, GitTags};

pub fn run(git_state: &GitState, git_tags: &GitTags, plan: u32) -> i32 {
    // Built once, not once per use: every line costs a git subprocess, and
    // is_branch_pushed_and_not_ahead is the one read query that writes a
    // diagnostic to stderr (plan-107), so evaluating twice would double it.
    let lines = report(git_state, git_tags, plan);
    let passed = lines.last().is_some_and(|l| l == "PASS");
    for line in lines {
        println!("{line}");
    }
    if passed {
        0
    } else {
        1
    }
}

/// The lines preflight prints, in order. Returned rather than printed so the
/// fail-fast/warn-accumulate ordering is testable without capturing stdout.
fn report(git_state: &GitState, git_tags: &GitTags, plan: u32) -> Vec<String> {
    if !git_state.is_clean() {
        return vec![
            "FAIL: working tree has uncommitted changes (git status --porcelain)".to_string()
        ];
    }

    let version_tag = git_tags.get_plan_version(plan);
    if !git_state.tag_exists_locally(&version_tag) {
        return vec![format!("FAIL: version tag {version_tag} not found locally")];
    }

    let mut lines = Vec::new();
    if !git_state.is_branch_pushed_and_not_ahead() {
        lines.push("WARN: branch is not pushed or HEAD is ahead of upstream".to_string());
    }
    if !git_state.tag_exists_on_remote(&version_tag) {
        lines.push(format!("WARN: version tag {version_tag} not found on remote"));
    }
    lines.push("PASS".to_string());
    lines
}
